### ISSUE TITLE
[Master][PDDF] PDDF framework FAN/PSU/xcvr drivers are incompatible with kernel 6.1

### ISSUE THREAD

**Issue by qnos** [sonic-net/sonic-buildimage#18677]
#### Description

[Master][PDDF] PDDF framework FAN/PSU/xcvr drivers are incompatible with kernel 6.1

PDDF FAN/PSU/xcvr drivers are using hwmon_device_register_with_info to register hwmon device, the chip variable are mandatory for linux kernel 6.1 but optional for kernel 5.10.
NULL pointer was passed in PDDF driver code for master branch, so it would fail and return EINVAL. 

```
$ rg -n "hwmon_device_register_with_info"
i2c/modules/xcvr/driver/pddf_xcvr_driver.c
162:    data->xdev = hwmon_device_register_with_info(&client->dev, client->name, NULL, NULL, NULL);

i2c/modules/fan/driver/pddf_fan_driver.c
519:    data->hwmon_dev = hwmon_device_register_with_info(&client->dev, client->name, NULL, NULL, NULL);

i2c/modules/psu/driver/pddf_psu_driver.c
246:    data->hwmon_dev = hwmon_device_register_with_info(&client->dev, client->name, NULL, NULL, NULL);
```

Linux Kernel 6.1:

```
/**                                                                                                                                                                                                  
 * hwmon_device_register_with_info - register w/ hwmon                                                                                                                                               
 * @dev: the parent device (mandatory)                                                                                                                                                               
 * @name: hwmon name attribute (mandatory)                                                                                                                                                           
 * @drvdata: driver data to attach to created device (optional)                                                                                                                                      
 * @chip: pointer to hwmon chip information (mandatory)                                                                                                                                              
 * @extra_groups: pointer to list of additional non-standard attribute groups                                                                                                                        
 *  (optional)                                                                                                                                                                                       
 *                                                                                                                                                                                                   
 * hwmon_device_unregister() must be called when the device is no                                                                                                                                    
 * longer needed.                                                                                                                                                                                    
 *                                                                                                                                                                                                   
 * Returns the pointer to the new device.                                                                                                                                                            
 */                                                                                                                                                                                                  
struct device *                                                                                                                                                                                      
hwmon_device_register_with_info(struct device *dev, const char *name,                                                                                                                                
                void *drvdata,                                                                                                                                                                       
                const struct hwmon_chip_info *chip,                                                                                                                                                  
                const struct attribute_group **extra_groups)                                                                                                                                         
{                                                                                                                                                                                                    
    if (!dev || !name || !chip)                                                                                                                                                                      
        return ERR_PTR(-EINVAL);                                                                                                                                                                     
                                                                                                                                                                                                     
    if (!chip->ops || !chip->ops->is_visible || !chip->info)                                                                                                                                         
        return ERR_PTR(-EINVAL);                                                                                                                                                                     
                                                                                                                                                                                                     
    return __hwmon_device_register(dev, name, drvdata, chip, extra_groups);                                                                                                                          
}                                                                                                                                                                                                    
EXPORT_SYMBOL_GPL(hwmon_device_register_with_info);  
```

In Linux Kernel 5.10, chip variable is optional:

```
/**                                                                                                                                                                                                  
 * hwmon_device_register_with_info - register w/ hwmon                                                                                                                                               
 * @dev: the parent device                                                                                                                                                                           
 * @name: hwmon name attribute                                                                                                                                                                       
 * @drvdata: driver data to attach to created device                                                                                                                                                 
 * @chip: pointer to hwmon chip information                                                                                                                                                          
 * @extra_groups: pointer to list of additional non-standard attribute groups                                                                                                                        
 *                                                                                                                                                                                                   
 * hwmon_device_unregister() must be called when the device is no                                                                                                                                    
 * longer needed.                                                                                                                                                                                    
 *                                                                                                                                                                                                   
 * Returns the pointer to the new device.                                                                                                                                                            
 */                                                                                                                                                                                                  
struct device *                                                                                                                                                                                      
hwmon_device_register_with_info(struct device *dev, const char *name,                                                                                                                                
                void *drvdata,                                                                                                                                                                       
                const struct hwmon_chip_info *chip,                                                                                                                                                  
                const struct attribute_group **extra_groups)                                                                                                                                         
{                                                                                                                                                                                                    
    if (!name)                                                                                                                                                                                       
        return ERR_PTR(-EINVAL);                                                                                                                                                                     
                                                                                                                                                                                                     
    if (chip && (!chip->ops || !chip->ops->is_visible || !chip->info))                                                                                                                               
        return ERR_PTR(-EINVAL);                                                                                                                                                                     
                                                                                                                                                                                                     
    if (chip && !dev)                                                                                                                                                                                
        return ERR_PTR(-EINVAL);                                                                                                                                                                     
                                                                                                                                                                                                     
    return __hwmon_device_register(dev, name, drvdata, chip, extra_groups);                                                                                                                          
}                                                                                                                                                                                                    
EXPORT_SYMBOL_GPL(hwmon_device_register_with_info); 
```

#### Steps to reproduce the issue:

Any platforms based on PDDF framework and using PDDF FAN/PSU/xcvr drivers in master branch build.

#### Describe the results you received:

```
Apr 11 03:14:10.617591 sonic INFO kernel: [   15.955742] i2c i2c-104: new_device: Instantiated device 24lc64t at 0x57
Apr 11 03:14:10.899618 sonic INFO kernel: [   16.228735] pddf_fan 105-0016: chip found
Apr 11 03:14:10.899641 sonic ERR kernel: [   16.228799] pddf_fan_probe: hwmon_device_register_with_info failed: dev name fan_cpld, ret '-22'.
Apr 11 03:14:10.899643 sonic WARNING kernel: [   16.237851] PDDF_FAN    pddf_fan_probe: Freed all the memory allocated for attributes
Apr 11 03:14:10.899644 sonic WARNING kernel: [   16.237859] pddf_fan: probe of 105-0016 failed with error -22
Apr 11 03:14:10.899645 sonic WARNING kernel: [   16.237870] PDDF_FAN    Created a FAN-CTRL client: 0x00000000af7a5f7f
Apr 11 03:14:10.899647 sonic WARNING kernel: [   16.237875] PDDF_CLIENT add_device_table: Adding ptr 0x00000000af7a5f7f to the hash table
Apr 11 03:14:10.905609 sonic WARNING kernel: [   16.242818] PDDF_CPLD   Created CPLD_BASEBOARD client: 0x00000000cded6f67
Apr 11 03:14:10.905621 sonic WARNING kernel: [   16.242829] PDDF_CLIENT add_device_table: Adding ptr 0x00000000cded6f67 to the hash table
Apr 11 03:14:10.913589 sonic INFO kernel: [   16.249915] i2c i2c-106: Added multiplexed i2c bus 42
Apr 11 03:14:10.913607 sonic INFO kernel: [   16.250108] i2c i2c-106: Added multiplexed i2c bus 43
Apr 11 03:14:10.913609 sonic INFO kernel: [   16.250315] i2c i2c-106: Added multiplexed i2c bus 44
```

#### Describe the results you expected:
 
No failure when setting up the devices:

```
2024-04-08T13:08:10.506996+00:00 sonic kernel: [   17.167807] i2c i2c-104: new_device: Instantiated device 24lc64t at 0x57
2024-04-08T13:08:10.506996+00:00 sonic kernel: [   17.167807] i2c i2c-104: new_device: Instantiated device 24lc64t at 0x57
2024-04-08T13:08:10.695014+00:00 sonic kernel: [   17.356648] pddf_fan 105-0016: chip found
2024-04-08T13:08:10.695035+00:00 sonic kernel: [   17.356727] pddf_fan 105-0016: hwmon35: fan 'fan_cpld'
2024-04-08T13:08:10.695037+00:00 sonic kernel: [   17.356752] PDDF_FAN  Created a FAN-CTRL client: 0x0000000021add19c
2024-04-08T13:08:10.695038+00:00 sonic kernel: [   17.356756] PDDF_CLIENT       add_device_table: Adding ptr 0x0000000021add19c to the hash table
2024-04-08T13:08:10.695014+00:00 sonic kernel: [   17.356648] pddf_fan 105-0016: chip found
2024-04-08T13:08:10.695035+00:00 sonic kernel: [   17.356727] pddf_fan 105-0016: hwmon35: fan 'fan_cpld'
2024-04-08T13:08:10.695037+00:00 sonic kernel: [   17.356752] PDDF_FAN  Created a FAN-CTRL client: 0x0000000021add19c
2024-04-08T13:08:10.695038+00:00 sonic kernel: [   17.356756] PDDF_CLIENT       add_device_table: Adding ptr 0x0000000021add19c to the hash table
2024-04-08T13:08:10.698976+00:00 sonic kernel: [   17.360206] PDDF_CPLD Created CPLD_BASEBOARD client: 0x000000009fb1577f
2024-04-08T13:08:10.698987+00:00 sonic kernel: [   17.360214] PDDF_CLIENT       add_device_table: Adding ptr 0x000000009fb1577f to the hash table
2024-04-08T13:08:10.698976+00:00 sonic kernel: [   17.360206] PDDF_CPLD Created CPLD_BASEBOARD client: 0x000000009fb1577f
2024-04-08T13:08:10.698987+00:00 sonic kernel: [   17.360214] PDDF_CLIENT       add_device_table: Adding ptr 0x000000009fb1577f to the hash table
2024-04-08T13:08:10.703027+00:00 sonic kernel: [   17.365147] i2c i2c-106: Added multiplexed i2c bus 42
```

#### Output of `show version`:

Issue in master branching only, caused by linux kernel hwmon API implementation changes.

#### Output of `show techsupport`:



#### Additional information you deem important (e.g. issue happens only occasionally):

<!--
     Also attach debug file produced by `sudo generate_dump`
-->

**Comment by qnos** [comment]
Some possible solutions:
1. Still uses `hwmon_device_register_with_info` API but create a stub device and pass in the device info.
2. Switch to `hwmon_device_register_with_groups` API, but this requires to refactor the PSU/FAN hwmon properties implementation.

@FuzailBrcm Any opinions from you?

**Comment by FuzailBrcm** [comment]
It is already addressed via https://github.com/sonic-net/sonic-buildimage/pull/18537

**Comment by qnos** [comment]
> It is already addressed via #18537

Thanks for the information. Duplicated with https://github.com/sonic-net/sonic-buildimage/pull/18537, close it.
